test/extension/handlers/topologymutation/handler.go

```diff
@@ -19,70 +19,239 @@ package topologymutation
 
 import (
 	"context"
+	"fmt"
 	"strconv"
+	"strings"
 
+	"github.com/blang/semver"
+	"github.com/pkg/errors"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/runtime/serializer"
+	intstrutil "k8s.io/apimachinery/pkg/util/intstr"
 	ctrl "sigs.k8s.io/controller-runtime"
 
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
 	runtimehooksv1 "sigs.k8s.io/cluster-api/exp/runtime/hooks/api/v1alpha1"
 	patchvariables "sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches/variables"
 	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util/version"
 )
 
 // NewHandler returns a new topology mutation Handler.
 func NewHandler(scheme *runtime.Scheme) *Handler {
 	return &Handler{
+		// Add the apiGroups being handled to the decoder
 		decoder: serializer.NewCodecFactory(scheme).UniversalDecoder(
 			infrav1.GroupVersion,
+			controlplanev1.GroupVersion,
+			bootstrapv1.GroupVersion,
 		),
 	}
 }
 
+var (
+	cgroupDriverCgroupfs   = "cgroupfs"
+	versionCeilingForPatch = semver.Version{Major: 1, Minor: 24}
+)
+
 // Handler is a topology mutation handler.
 type Handler struct {
 	decoder runtime.Decoder
 }
 
-// GeneratePatches returns a function that generates patches for the given request.
+// GeneratePatches generates patches for the given request.
 func (h *Handler) GeneratePatches(ctx context.Context, req *runtimehooksv1.GeneratePatchesRequest, resp *runtimehooksv1.GeneratePatchesResponse) {
 	log := ctrl.LoggerFrom(ctx)
-	log.Info("GeneratePatches called")
 
+	log.Info("GeneratePatches called")
 	walkTemplates(h.decoder, req, resp, func(obj runtime.Object, variables map[string]apiextensionsv1.JSON) error {
-		if dockerClusterTemplate, ok := obj.(*infrav1.DockerClusterTemplate); ok {
-			if err := patchDockerClusterTemplate(dockerClusterTemplate, variables); err != nil {
-				return err
+		log.Info("Patching template", "template", obj.GetObjectKind().GroupVersionKind().String())
+
+		switch typedObj := obj.(type) {
```

Comment:
```suggestion
		switch obj := obj.(type) {
```
nit: totally up to you. 

I usually saw this done like this. I assume to shadow the var to ensure that the untyped var is not used inside of the switch

Reply:
>nit: totally up to you.

😆  I think this one is up to you! `typedObj` is from your suggestion to turn this into a switch statement, but I'm happy whichever way, I think it's pretty inconsequential.

Reply:
Ah. I think I missed this when making the suggestion. I would go with my latest suggestion (last one - I promise :))